crates/external_services/src/grpc_client/health_check_client.rs

```
@@ -0,0 +1,112 @@
+    clippy::use_self
+)]
+pub mod health_check {
+    tonic::include_proto!("grpc.health.v1");
+}
+
+use super::{DynamicRoutingClientConfig, GrpcClientSettings};
+
+/// Result type for Dynamic Routing
+pub type HealthCheckResult<T> = CustomResult<T, HealthCheckError>;
+/// Dynamic Routing Errors
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum HealthCheckError {
+    /// The required input is missing
+    #[error("Missing Host and Port for building the Health check connection")]
+    MissingFields,
+    /// Error from gRPC Server
+    #[error("Error from gRPC Server : {0}")]
+    ConnectionError(String),
+    /// status is invalid
+    #[error("Invalid Status from server")]
+    InvalidStatus,
+}
+
+/// Health Check Client type
+#[derive(Debug, Clone)]
+pub struct HealthCheckClient;
+
+impl HealthCheckClient {
+    /// Perform health check for all services involved
+    pub async fn perform_health_check(
+        &self,
+        config: &GrpcClientSettings,
+    ) -> HealthCheckResult<HealthCheckMap> {
+        let dynamic_routing_config = &config.dynamic_routing_client;
+        let connection = match dynamic_routing_config {
+            DynamicRoutingClientConfig::Enabled {
+                host,
+                port,
+                service,
+            } => Some((host.clone(), *port, service.clone())),
+            _ => None,
+        };
+
+        // SAFETY : This is a safe cast as there exists a valid
+        // integer value for this variant
+        #[allow(clippy::as_conversions)]
+        let expected_status = ServingStatus::Serving as i32;
+
+        let mut service_map = HealthCheckMap::new();
+
+        let health_check_succeed = connection
+            .async_map(|conn| self.get_response_from_grpc_service(conn.0, conn.1, conn.2))
+            .await
+            .transpose()
+            .change_context(HealthCheckError::ConnectionError(
+                "error calling dynamic routing service".to_string(),
+            ))
+            .ok()
```

Error is dropped after ok and is not propagated or logged?